### ISSUE TITLE
RError: Cannot determine how to set picked for lineid[2]! - GERS unfill causes error

### ISSUE THREAD

**Issue by kbower** [kbower/tickettest1#1417]
_Imported from trac ticket 1417.
Created by **kbower**
Opened in trac: Jan 31, 2012, 9:56:42 AM
Last modified in trac: Feb 10, 2012, 3:24:25 PM_
***
_by kbower:_
RError: Cannot determine how to set picked for lineid[[2]](../commit/6f1081b3eac5ffa7bdb709e3177addf50c049987)! order.type[SAL] voided[N] fillsiteid[None] filllocationid[None] filldate[2012-01-29] originalinventorycost[None] product.stockeditem[Y]

```
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] File '/home/rarch/trunk/src/appserver/pylotengine/controllers/customerorder.py', line 22 in save
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   response = self._save(*args, **params)
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] File '/home/rarch/trunk/src/appserver/pylotengine/controllers/dbentity.py', line 329 in _save
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   return self._save_obj(obj, **kwargs)
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] File '/home/rarch/trunk/src/appserver/pylotengine/controllers/dbentity.py', line 336 in _save_obj
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   merged_obj = self.modelClass.save(passed_obj, controller_validate)
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] File '/home/rarch/trunk/src/appserver/pylotengine/model/objects.py', line 457 in save
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   DBSession.flush()
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] File '/home/rarch/tg2env/lib/python2.6/site-packages/SQLAlchemy-0.6.4.4kbdev-py2.6-linux-x86_64.egg/sqlalchemy/orm/scoping.py', line 132 in do
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   return getattr(self.registry(), name)(*args, **kwargs)
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] File '/home/rarch/tg2env/lib/python2.6/site-packages/SQLAlchemy-0.6.4.4kbdev-py2.6-linux-x86_64.egg/sqlalchemy/orm/session.py', line 1367 in flush
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   self._flush(objects)
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] File '/home/rarch/tg2env/lib/python2.6/site-packages/SQLAlchemy-0.6.4.4kbdev-py2.6-linux-x86_64.egg/sqlalchemy/orm/session.py', line 1448 in _flush
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   flush_context.execute()
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] File '/home/rarch/tg2env/lib/python2.6/site-packages/SQLAlchemy-0.6.4.4kbdev-py2.6-linux-x86_64.egg/sqlalchemy/orm/unitofwork.py', line 299 in execute
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   rec.execute(self)
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] File '/home/rarch/tg2env/lib/python2.6/site-packages/SQLAlchemy-0.6.4.4kbdev-py2.6-linux-x86_64.egg/sqlalchemy/orm/unitofwork.py', line 443 in execute
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   uow
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] File '/home/rarch/tg2env/lib/python2.6/site-packages/SQLAlchemy-0.6.4.4kbdev-py2.6-linux-x86_64.egg/sqlalchemy/orm/mapper.py', line 1618 in _save_obj
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   mapper, conn, state.obj())
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] File '/home/rarch/tg2env/lib/python2.6/site-packages/SQLAlchemy-0.6.4.4kbdev-py2.6-linux-x86_64.egg/sqlalchemy/orm/util.py', line 241 in _do
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   ret = getattr(ext, method)(*args, **kwargs)
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] File '/home/rarch/trunk/src/appserver/pylotengine/model/mappings.py', line 294 in before_update
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   self._set_picked_flag(instance)
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] File '/home/rarch/trunk/src/appserver/pylotengine/model/mappings.py', line 317 in _set_picked_flag
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   (l.lineid, order.type, l.voided, l.fillsiteid, l.filllocationid, l.filldate, l.originalinventorycost, l.product.stockeditem))
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] RError: Cannot determine how to set picked for lineid[5]! order.type[SAL] voided[N] fillsiteid[None] filllocationid[None] filldate[2012-01-28] originalinventorycost[None] product.stockeditem[Y]
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] CGI Variables
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] -------------
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   CONTENT_LENGTH: '-1'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   CONTENT_TYPE: 'application/x-www-form-urlencoded'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   DOCUMENT_ROOT: '/home/rarch/trunk/src/appserver/wsgi-config/flex'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   GATEWAY_INTERFACE: 'CGI/1.1'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   HTTPS: '1'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   HTTP_ACCEPT: '*/*'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   HTTP_ACCEPT_LANGUAGE: 'en-US'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   HTTP_AUTHORIZATION: 'Basic TUhPQ0hBOkFDVDNLSg=='
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   HTTP_CACHE_CONTROL: 'no-cache'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   HTTP_CONNECTION: 'Keep-Alive'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   HTTP_COOKIE: 'pylotengine=ae5d3fe81b51bbf46505e942fba4bfbe28a187c2cdd5c00ed808e9ae430654449708dc76'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   HTTP_HOST: 'kanes.retailarchitects.com'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   HTTP_REFERER: 'https://kanes.retailarchitects.com/SalesPylot.swf/[[DYNAMIC]]/6'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   HTTP_USER_AGENT: 'Mozilla/4.0 (compatible; MSIE 6.0; Windows NT 5.0; .NET CLR 1.1.4322; .NET CLR 2.0.50727)'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   HTTP_X_FLASH_VERSION: '10,1,102,64'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   PATH_INFO: '/customerorder/save'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   PATH_TRANSLATED: '/home/rarch/trunk/src/appserver/wsgi-config/flex/customerorder/save'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   REMOTE_ADDR: '173.221.136.228'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   REMOTE_PORT: '54233'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   REMOTE_USER: 'MHOCHA'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   REQUEST_METHOD: 'POST'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   REQUEST_URI: '/tg/customerorder/save'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   SCRIPT_FILENAME: '/home/rarch/trunk/src/appserver/wsgi-config/wsgi-deployment.py'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   SCRIPT_NAME: '/tg'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   SERVER_ADDR: '173.204.243.250'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   SERVER_ADMIN: 'root@localhost'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   SERVER_NAME: 'kanes.retailarchitects.com'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   SERVER_PORT: '443'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   SERVER_PROTOCOL: 'HTTP/1.0'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   SERVER_SIGNATURE: '<address>Apache/2.2.3 (CentOS) Server at kanes.retailarchitects.com Port 443</address>\\n'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   SERVER_SOFTWARE: 'Apache/2.2.3 (CentOS)'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] WSGI Variables
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] --------------
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   application: <function wrapper at 0x2af0ed19b848>
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   beaker.cache: <beaker.cache.CacheManager object at 0x2af0ed19d590>
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   beaker.get_session: <bound method SessionMiddleware._get_session of <beaker.middleware.SessionMiddleware object at 0x2af0ed19d510>>
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   beaker.session: {'_accessed_time': 1327855543.6795771}
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   downgrade-1.0: '1'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   force-response-1.0: '1'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   mod_wsgi.application_group: ''
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   mod_wsgi.callable_object: 'application'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   mod_wsgi.handler_script: ''
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   mod_wsgi.input_chunked: '0'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   mod_wsgi.listener_host: ''
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   mod_wsgi.listener_port: '443'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   mod_wsgi.process_group: 'rarch'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   mod_wsgi.request_handler: 'wsgi-script'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   mod_wsgi.script_reloading: '1'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   mod_wsgi.version: (3, 3)
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   nokeepalive: '1'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   paste.registry: <paste.registry.Registry object at 0x2aaaac58cf50>
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   paste.throw_errors: True
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   pylons.action_method: <bound method RootController.routes_placeholder of <pylotengine.controllers.root.RootController object at 0x2aaaac5a7250>>
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   pylons.controller: <pylotengine.controllers.root.RootController object at 0x2aaaac5a7250>
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   pylons.environ_config: {'session': 'beaker.session', 'cache': 'beaker.cache'}
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   pylons.pylons: <pylons.util.PylonsContext object at 0x2aaaac5a7190>
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   pylons.routes_dict: {'url': u'customerorder/save', 'action': u'routes_placeholder', 'controller': u'root'}
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   pylons.status_code_redirect: None
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   pylotengine.flex_detected: True
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   pylotengine.flex_revision: '1850'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   repoze.tm.active: True
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   repoze.what.adapters: {'groups': {'sql_auth': <repoze.what.plugins.sql.adapters.SqlGroupsAdapter object at 0x2af0ed19d790>}, 'permissions': {'sql_auth': <repoze.what.plugins.sql.adapters.SqlPermissionsAdapter object at 0x2af0ed19d810>}}
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   repoze.what.credentials: {'repoze.what.userid': 'MHOCHA', 'groups': (u'sales',), 'permissions': (u'ordernew', u'orderpay', u'productsalesinv', u'ordereditownlimited', u'kanprintinvoiceanyway', u'orderquery', u'orderedit', u'ordertypesale', u'ordercashier')}
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   repoze.who.identity: <repoze.who identity (hidden, dict-like) at 46912610433152>
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   repoze.who.logger: <logging.Logger instance at 0x2af0e8e6aa70>
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   repoze.who.plugins: {'form': <FriendlyFormPlugin 47214258411728>, 'main_identifier': <FriendlyFormPlugin 47214258411728>, 'authorization_md': <repoze.what.middleware.AuthorizationMetadata object at 0x2af0ed19d950>, 'api': <BasicAuthPlugin 47214212941712>, 'cookie': <AuthTktCookiePlugin 47214258411664>, 'sqlauth': <repoze.who.plugins.sa.SQLAlchemyAuthenticatorPlugin object at 0x2af0ed19d850>, 'sql_user_md': <repoze.who.plugins.sa.SQLAlchemyUserMDPlugin object at 0x2af0ed19d910>}
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   routes.route: <routes.route.Route object at 0x2af0ea79b510>
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   routes.url: <routes.util.URLGenerator object at 0x2aaaac587fd0>
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   serialized_properties: {<class 'pylotengine.model.objects.CustomerOrder'>: [('currentorderid', True)]}
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   ssl-unclean-shutdown: '1'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   toscawidgets.framework: <tw.mods.base.HostFramework object at 0x2af0ed19d610>
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   toscawidgets.prefix: '/toscawidgets'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   webob._parsed_post_vars: (MultiDict([('CustomerOrder', '{"restricted_attributes":null,"customer":{"restricted_attributes":null,"firstname":"JUDITH ","lastname":"BOWMAN","address2":null,"companyname":null,"address1":"17031 HIGHWAY 301 LOT143","zipcode":"33523","city":"Dade City","email":null,"recordgrouping":null,"phonenumber":"352-567-1363","type":"I","taxexemptionid":null,"fullname":"JUDITH  BOWMAN","picture_url":"resources/images/blankuser.jpg","warnings":[],"businessphone":null,"accountopendate":"2012-01-28 00:00:00","state":"FL","middlename":null,"creatingCustomer":false,"taxexemptionnumber":null,"country":null,"extension":null,"customerid":"68466"},"orders":[{"restricted_attributes":{"percentofsale1":["{NoPermission}",[["ORDERS - Ability to change the salepersons or percentage of sale for an order."]]],"salesperson1":["{NoPermission}",[["ORDERS - Ability to change the salepersons or percentage of sale for an order."]]],"percentofsale2":["{NoPermission}",[["ORDERS - Ability to change th...t([('Cu...}')])>)
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   webob._parsed_query_vars: (MultiDict([]), '')
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   webob.adhoc_attrs: {'errors': 'ignore', 'language': 'en-us', 'start_response': <function repl_start_response at 0x2aaaabaf06e0>, 'response_type': None, 'response_ext': None, 'identity': <repoze.who identity (hidden, dict-like) at 46912610433152>}
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   wsgi process: 'Multi process AND threads (?)'
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   wsgi.file_wrapper: <built-in method file_wrapper of mod_wsgi.Adapter object at 0x2aaab0c79eb8>
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   wsgi.version: (1, 1)
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   wsgiorg.routing_args: (<routes.util.URLGenerator object at 0x2aaaac587fd0>, {'url': u'customerorder/save', 'action': u'routes_placeholder', 'controller': u'root'})
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   <symbol 'SAVING>: True
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228]   <symbol 'FOR_CLIENT_WRITE>: True
[Sun Jan 29 11:45:47 2012] [error] [client 173.221.136.228] ------------------------------------------------------------
```
***
* kbower attached [None](https://trac.retailarchitects.com/trac/attachment/ticket/1417/None) on Feb 6, 2012, 8:43:43 PM

**Comment by kbower** [comment]
_**kbower** trac comment on Jan 31, 2012, 10:18:17 AM:_

(Changeset [[2247]](../commit/7b42431a39814046f01d799e8b72675ebaf4d162))
 * better debug error, re #1417

**Comment by kbower** [comment]
_**kbower** trac comment on Jan 31, 2012, 10:24:08 AM:_

Wayne,
    Can you think of what may cause a SAL line to have NULL store_cd and NULL loc_cd, but have a date in FILL_DT?

This causes an exception sometimes: *RError: Cannot determine how to set picked for lineid[[2]](../commit/6f1081b3eac5ffa7bdb709e3177addf50c049987)! order.type[SAL] voided[N] fillsiteid[None] filllocationid[None] filldate[2012-01-29] originalinventorycost[None] product.stockeditem[Y]*

**Comment by kbower** [comment]
_**mpisarski** trac comment on Feb 3, 2012, 2:28:59 PM:_

**Mariam wrote:**
Order: 0203216MI22
Hello,
I encountered this error while trying to fill from this store.
Thank you,
Mariam
 

**Michael wrote:**
Mariam,

Looking from the order data and the logged System comments, it looks like this order was unfilled from GERS but only the fill site and location were unset, but the fill date was left. In our system we check to make sure that all three (fill site, fill location and fill date) are blank before we fill the line.
Is this how GERS normally works? Can you run a test in Train to see if GERS always leaves the Fill Date in place if you unfill the lines?

Thanks,

Michael

**Comment by kbower** [comment]
_**kbower** trac comment on Feb 6, 2012, 11:17:53 AM:_

Wayne,
    Given that sometimes FILL_DT is NOT NULL after gers unfills a line, how do these requirements change (particularly for 'Z' and 'X') regarding setting SO_LN.PICKED?

**Comment by kbower** [comment]
_**kbower** trac comment on Feb 6, 2012, 12:52:05 PM:_

I need to ask different questions:

What should the SO_LN.PICKED indicator be set to:
 * When STORE_CD, LOC_CD are null *and FILL_DT is not null* and VOID_FLAG = 'N' AND ITM.INVENTORY = 'Y'
 * When STORE_CD, LOC_CD are null *and FILL_DT is not null *and VOID_FLAG = 'N' AND ITM.INVENTORY = 'N'

?

**Comment by kbower** [comment]
_**kbower** trac comment on Feb 6, 2012, 8:43:43 PM:_

[attachment:"None"]

**Comment by kbower** [comment]
_**kbower** trac comment on Feb 6, 2012, 8:45:46 PM:_

Kent,

 * When STORE_CD, LOC_CD are null and FILL_DT is not null and VOID_FLAG = 'N' AND ITM.INVENTORY = 'Y'
   * SO_LN.PICKED = ‘Z’


 * When STORE_CD, LOC_CD are null and FILL_DT is not null and VOID_FLAG = 'N' AND ITM.INVENTORY = 'N'
   * SO_LN.PICKED = ‘X’

The second scenario should not happen since if an item is not an inventory item it should never get a fill date however we have had two occurrences of this in the last two years, none of which was in the last year.

Thanks,

Wayne Liburd

**Comment by kbower** [comment]
_**kbower** trac comment on Feb 6, 2012, 8:46:20 PM:_

Since the second scenario shouldn't happen, make sure that one continues to raise an error, Michael.

**Comment by kbower** [comment]
_**mpisarski** trac comment on Feb 10, 2012, 3:24:25 PM:_

(Changeset [[2289]](../commit/8625bfd159ac1aa27639aee303a2d304f053ec8d))
 * fixes #1417 - RError: Cannot determine how to set picked for lineid - when items are unfilled from GERS - Loft could not re-fill those items.

**Comment by kbower** [comment]
_**mpisarski** trac status change on Feb 10, 2012, 3:24:25 PM:_ **closed**
